### PR TITLE
Configure authentication at Build time

### PR DESCRIPTION
**Pull request by rmanibus**
solve #797 

Many thanks for submitting your Pull Request :heart:!

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Your code is properly formatted according to [our code style](CONTRIBUTING.md#code-style)
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] Subject`
- [x] Pull Request contains link to the issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-quarkus2` to backport the original PR to the `quarkus2` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

**Comment by rmanibus** [comment]
there is an issue with the fact that I made the OIDC Filter extension optional. I provided a temporary fix by removing the dependency to AbstractOidcClientRequestReactiveFilter and AbstractOidcClientRequestFilter (see the latest commit),
but there is probably a better way of doing this.

before the fix,
with rest easy classic the build hung with the following error:

`2024-09-14 19:00:36,019 WARN  [io.qua.arc.pro.BeanArchives] (build-17) Failed to index io.quarkus.oidc.client.reactive.filter.runtime.AbstractOidcClientRequestReactiveFilter: Class does not exist in ClassLoader QuarkusClassLoader:Deployment Class Loader: TEST for BeanValidationTest (QuarkusTest)@4bbf38b8`

with rest easy reactive the build hung with the following error:

`2024-09-14 18:54:22,716 WARN  [io.qua.arc.pro.BeanArchives] (build-12) Failed to index io.quarkus.oidc.client.filter.runtime.AbstractOidcClientRequestFilter: Class does not exist in ClassLoader QuarkusClassLoader:Deployment Class Loader: TEST for BeanValidationTest (QuarkusTest)@1c30cb85`

File: client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/GeneratorProcessor.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+import jakarta.enterprise.inject.Instance;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.ClassType;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.ParameterizedType;
+
+import io.quarkiverse.openapi.generator.*;
+import io.quarkiverse.openapi.generator.markers.*;
+import io.quarkiverse.openapi.generator.providers.*;
```
Comment (mcruzdev) [inline client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/GeneratorProcessor.java:18]: Replace with single class imports `*`

**Review by mcruzdev — APPROVED** [review]
Hi @rmanibus, amazing Job! I just added a simple comment

**Comment by rmanibus** [comment]
@mcruzdev I fixed the imports !

**Comment by mcruzdev** [comment]
> @mcruzdev I fixed the imports !

I learned a lot, thank you for this pull request!

**Comment by rmanibus** [comment]
@mcruzdev you are very welcome ! Any chance to get this approved again ? The force push dropped the review

**Review by mcruzdev — APPROVED** [review]

**Review by hbelmiro — COMMENTED** [review]
Thank you for this contribution @rmanibus. Nicely done!
I just left some aesthetical nits.

**Comment by rmanibus** [comment]
@hbelmiro thanks a lot ! It is done !

**Review by hbelmiro — APPROVED** [review]
Thank you @rmanibus.

File: client/runtime/src/main/java/io/quarkiverse/openapi/generator/providers/CompositeAuthenticationProvider.java [hunk 2]
```diff
@@ -17,20 +16,20 @@
...
-public abstract class AbstractCompositeAuthenticationProvider implements ClientRequestFilter {
+public class CompositeAuthenticationProvider implements ClientRequestFilter {
 
-    private final List<AuthProvider> authProviders = new ArrayList<>();
+    private final List<AuthProvider> authProviders;
 
-    public final void addAuthenticationProvider(final AuthProvider authProvider) {
-        this.authProviders.add(authProvider);
+    public CompositeAuthenticationProvider(List<AuthProvider> authProviders) {
+        this.authProviders = List.copyOf(authProviders);
```
Comment (fjtirado) [inline client/runtime/src/main/java/io/quarkiverse/openapi/generator/providers/CompositeAuthenticationProvider.java:24]: Why do we need to copy this?

**Review by fjtirado — APPROVED** [review]

File: client/runtime/src/main/java/io/quarkiverse/openapi/generator/providers/CompositeAuthenticationProvider.java (hunk 2 above)
Comment (rmanibus) [reply client/runtime/src/main/java/io/quarkiverse/openapi/generator/providers/CompositeAuthenticationProvider.java:24]: I wanted it to be immutable

Comment (rmanibus) [reply client/runtime/src/main/java/io/quarkiverse/openapi/generator/providers/CompositeAuthenticationProvider.java:24]: there is a public getter on it

**Review by ricardozanini — APPROVED** [review]

**Comment by ricardozanini** [comment]
@wmedvede will also take a look since he made many changes on this code too.

**Comment by rmanibus** [comment]
Any updates about this ?

**Pull request merged**